### PR TITLE
[DD-269] Node.JS & Python Guided Tours

### PR DESCRIPTION
**Pull request by nramabad**
# Links

[PREVIEW](http://circleci-doc-preview.s3-website-us-east-1.amazonaws.com/lang-guided-tour-preview/?force-all) • [FIGMA](https://www.figma.com/file/HrgpUU2M3ED7OT0p0f99bH/Docs-Disco-Experiments?node-id=3760%3A7933) • [JIRA](https://circleci.atlassian.net/browse/DD-269)

# Description
Develop Node & Python guided tours so we can test if end users would prefer a guided flow over long articles to set up config files.

# Impact
Validate if users prefer guided tours. If they would, we can extend it to more languages and improve the user experience to set up config files. Will measure the below metrics in amplitude:

1. Time it takes new users to get a working config with an Article, vs the time it takes new users to get a working config with a Tutorial
2. How far the users get in the tutorial before they fall out

# Definition of Done
1. Display Node & Python guides on Docs home page; Design is [here](https://www.figma.com/file/HrgpUU2M3ED7OT0p0f99bH/Docs-Disco-Experiments?node-id=3760%3A7933)
2. The Node flow works as [here](https://www.figma.com/file/HrgpUU2M3ED7OT0p0f99bH/Docs-Disco-Experiments?node-id=4189%3A9032), the Python flow is [here](https://www.figma.com/file/HrgpUU2M3ED7OT0p0f99bH/Docs-Disco-Experiments?node-id=3878%3A8009).  
3. Refer here for tutorial content: [Node](https://circleci.atlassian.net/wiki/spaces/PEX/pages/6387826851/DD-146+Node+content+for+Guided+Tour) & [Python](https://circleci.atlassian.net/l/c/1zg9ECG6)
4. Deploy both guides together.
5. Redirect to the experiment from the same Docs link that the  https://circleci.com/docs/2.0/language-javascript/,  https://circleci.com/docs/2.0/language-python/  articles live on now.
6. Provide tracking of how far into the tour our readers make it.  We want to know at what step readers fell off.  Refer to this [conversation](https://circleci.slack.com/archives/C021NCSEM44/p1641931275155600) for more detail.

# Before/Control

## Homepage
https://circleci.com/docs/
https://circleci.com/docs/2.0/

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245199-b39d7cc0-5306-4e38-a0f1-b01591b3c0bf.png" alt="Homepage"></img>
</details>

## Node Guide
https://circleci.com/docs/2.0/language-javascript/

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245735-fa1b6701-5a8a-45c2-a6e3-bbd08c71b9d4.png" alt="Node.JS"></img>
</details>

## Python Guide
https://circleci.com/docs/2.0/language-python/

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245695-c72b888e-ff8a-4d74-9c8d-1c9b1888a99f.png" alt="Python"></img>
</details>

# After/Treatment

## Homepage
http://circleci-doc-preview.s3-website-us-east-1.amazonaws.com/lang-guided-tour-preview/?force-all
http://circleci-doc-preview.s3-website-us-east-1.amazonaws.com/lang-guided-tour-preview/2.0/?force-all

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245292-273fb0dd-be3d-47d0-a73e-18d1b2d6261a.png" alt="Homepage"></img>
</details>

## Node Guide
http://circleci-doc-preview.s3-website-us-east-1.amazonaws.com/lang-guided-tour-preview/2.0/language-javascript/?force-all

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245982-e0596894-82f2-42c6-872f-6693c688df26.png" alt="Node.JS"></img>
</details>

## Python Guide
http://circleci-doc-preview.s3-website-us-east-1.amazonaws.com/lang-guided-tour-preview/2.0/language-python/?force-all

<details>
<summary>SCREENSHOT</summary>
<image src="https://user-images.githubusercontent.com/42252054/150245944-69d61515-6855-4058-9c1f-2a68e8a22f6b.png" alt="Python"></img>
</details>

### CODE REVIEW

File: jekyll/_cci2/language-javascript-experimental.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+description: "Building and Testing with JavaScript and Node.js on CircleCI"
+categories: [language-guides]
+order: 5
+version:
+- Cloud
+- Server v3.x
+- Server v2.x
+---
+
+{% raw %}
```
Comment (ccgeller) [inline jekyll/_cci2/language-javascript-experimental.md:14]: Why does this need to be wrapped in `raw` tags?

File: jekyll/_cci2/language-javascript-experimental.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
+---
+layout: classic-docs
```
Comment (ccgeller) [inline jekyll/_cci2/language-javascript-experimental.md:2]: classic-docs-experimental?

File: src/styles/main.scss [hunk 3]
```diff
@@ -754,3 +754,5 @@ Fixing styling for the code example tables in the "Migrating From..." pages
 #tooltip-time[data-popper-placement^=bottom] > #arrow {
   top: -4px;
 }
+
+@import 'language-guide';
```
Comment (ccgeller) [inline src/styles/main.scss:758]: suggestion nit: put imports at the top of document

File: jekyll/_cci2/language-javascript-experimental.md (hunk 2 above)
Comment (nramabad) [reply jekyll/_cci2/language-javascript-experimental.md:2]: that would create an infinite loop

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (nramabad) [reply jekyll/_cci2/language-javascript-experimental.md:14]: it's a jekyll/liquid bug that I encountered and this fixed it.

File: src/js/experiments/snippetFeedback.js [hunk 4]
```diff
@@ -52,7 +52,7 @@ export class SnippetFeedback {
       className: 'prompt',
       text: 'Was this helpful?',
     });
-    const slash = this._makeElement({ kind: 'span', text: ' / ' });
+    const slash = this._makeElement({ kind: 'span', text: ' /' });
```
Comment (nramabad) [inline src/js/experiments/snippetFeedback.js:55]: the extra space was bothering me.

File: src/js/experiments/snippetFeedback.test.js [hunk 5]
```diff
@@ -35,7 +35,7 @@ describe('Snippet Feedback class', () => {
       'Was this helpful?',
     );
     expect(sf.wasThisHelpfulContainer.children[1].textContent).toBe('Yes');
-    expect(sf.wasThisHelpfulContainer.children[2].textContent).toBe(' / ');
+    expect(sf.wasThisHelpfulContainer.children[2].textContent).toBe(' /');
```
Comment (nramabad) [inline src/js/experiments/snippetFeedback.test.js:38]: https://github.com/circleci/circleci-docs/pull/6111/files#r780582972
> the extra space was bothering me.

File: src/js/site/main.js [hunk 6]
```diff
@@ -260,6 +260,7 @@ $(document).ready(function () {
 
   // Give article headings direct links to anchors
   $('article h1, article h2, article h3, article h4, article h5, article h6')
+    .not('.card')
```
Comment (nramabad) [inline src/js/site/main.js:263]: need to change to '.card article h1, article h2, article h3, article h4, article h5, article h6'

Comment (nramabad) [reply src/js/site/main.js:263]: and add to correct place

File: src/js/experiments/languageGuides.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
+const displayElement = (el) => {
+  if (el) {
+    el.style.display = 'initial';
+  }
+};
+
+// https://app.optimizely.com/v2/projects/16812830475/experiments/20872380274/variations
+export default () =>
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_add_video_tab_to_docs_test',
```
Comment (nramabad) [inline src/js/experiments/languageGuides.js:10]: fix experiment key

File: jekyll/_layouts/classic-docs-experimental.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+layout: classic-docs
+section: cci2
+---
+<div class="control {{ page.categories | join: ' ' }}">
+    {{content}}
+</div>
+<div class="treatment no_toc_section {{ page.categories | join: ' ' }}">
+    {% if page.categories contains "language-guides" %}
+        {% capture lang_guide %}
+            {% include_relative language-{{page.short-title | downcase}}-experimental.md %}
```
Comment (nramabad) [inline jekyll/_layouts/classic-docs-experimental.html:11]: try removing indent to see if it fixes pre code bug / a remove {% raw %}

File: jekyll/_layouts/classic-docs-experimental.html [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        {% endcapture %}
+        {{ lang_guide | markdownify }}
+    {% endif %}
+</div>
+<style>#full-height{visibility:hidden}</style>
+<noscript class="no_toc_section">
+    {{content}}
```
Comment (nramabad) [inline jekyll/_layouts/classic-docs-experimental.html:18]: try replacing w simple style tag

File: jekyll/index.html [hunk 10]
```diff
@@ -15,6 +15,35 @@ <h1>Welcome to CircleCI Documentation</h1>
 <hr class="hidden-xs" />
```
Comment (nramabad) [inline jekyll/index.html:15]: repeat for /2.0/

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (nramabad) [reply jekyll/_cci2/language-javascript-experimental.md:14]: otherwise it interprets it as a code block oddly.

File: src/js/experiments/languageGuides.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.treatment',
+    attributes: {
+      // https://app.optimizely.com/v2/projects/16812830475/audiences/20680670160
+      windowWidth: window.innerWidth ?? 0,
+    },
+  }).then((variation) => {
```
Comment (nramabad) [inline src/js/experiments/languageGuides.js:20]: pull out logic into reusable utility functions

File: jekyll/_includes/snippets/language-guided-tour-cards.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+{::options parse_block_html="true" toc_levels="3..6" /}
+
+<div id="cards">
+<div class="card">
+
+{% include snippets/language-card.md lang=include.lang %}
+
+<div id="completion-time" class="card">
+<img alt="Completion Time" src="{{ site.baseurl }}/assets/img/docs/duration.svg">
+<span>50 min read</span>
```
Comment (nramabad) [inline jekyll/_includes/snippets/language-guided-tour-cards.md:10]: replace w/ variable once final copy/mock is discussed w/ @danbeere

Comment (nramabad) [reply jekyll/_includes/snippets/language-guided-tour-cards.md:10]: also considering refactoring these 3 lines into an reusable include

File: jekyll/_includes/no-script.html [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<noscript>
+    <style>.row,.control{display:initial}.loading-deferred{display:block}#full-height{visibility:visible}</style>
```
Comment (tshen815) [inline jekyll/_includes/no-script.html:2]: is this to ensure that these styles are set in the event that scripting is turned off in the browser?

File: jekyll/index.html [hunk 14]
```diff
@@ -15,6 +15,33 @@ <h1>Welcome to CircleCI Documentation</h1>
...
+        {% capture python-card %}
+          {% include snippets/language-card.md lang="Python" anchor="true" %}
+        {% endcapture %}
+        {{ python-card | markdownify }}
+      </div>
+    </a>
+  </div>
+  <div class="treatment col-xs-12">
+    <hr />
+  </div>
```
Comment (tshen815) [inline jekyll/index.html:44]: Could we put this in a seperate .html file and use an include tag, or would that be overkill since we are most likely only using this here for the home page?

File: jekyll/_includes/no-script.html (hunk 13 above)
Comment (nramabad) [reply jekyll/_includes/no-script.html:2]: Also as to why we use display none.

It helps with regards to CLS and generally avoiding jankiness until the experiment is loaded.

File: jekyll/index.html (hunk 14 above)
Comment (nramabad) [reply jekyll/index.html:44]: I can look into it. No guarantee since the 2.0 index page is an MD file. So making the include + variables work for both md and html has been a challenge thus far. There's hacks to make it work that I can see if I can employ.

File: src/js/experiments/languageGuides.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+// https://app.optimizely.com/v2/projects/16812830475/experiments/20872380274/variations
+export default () =>
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_add_video_tab_to_docs_test',
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.treatment',
+    attributes: {
+      // https://app.optimizely.com/v2/projects/16812830475/audiences/20680670160
+      windowWidth: window.innerWidth ?? 0,
+    },
```
Comment (rsylvian) [inline src/js/experiments/languageGuides.js:19]: Are we preventing people on mobile to see that experiment?

File: src/js/experiments/languageGuides.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (homepage) {
+      Array.prototype.forEach.call(elements, displayInitialElement);
+      homepage.style.display = 'initial';
+      Array.prototype.forEach.call(deferred, displayBlockElement);
+      return;
+    }
+
+    const element = elements[0];
+    if (!element) return;
```
Comment (rsylvian) [inline src/js/experiments/languageGuides.js:37]: Can you please add a comment in the code to explain what it is given than `element` is used everywhere after

File: src/js/experiments/relatedResources.js [hunk 17]
```diff
@@ -1,24 +1,22 @@
...
+  },
+}).then((variation) => {
+  if (variation === 'treatment') {
+    $('#related-resources').show();
+    // track clicks on each items
+    $('#related-resources li a').click(function () {
+      window.AnalyticsClient.trackAction('docs-related-resources-clicked', {
+        page: location.pathname,
+        link: $(this).attr('href'),
+        name: $(this).text(),
```
Comment (rsylvian) [inline src/js/experiments/relatedResources.js:18]: Unless it is related to that PR, can we make another PR for that, please?

File: src/js/experiments/surfaceFullExampleConfig.js [hunk 18]
```diff
@@ -1,17 +1,15 @@
...
+window.OptimizelyClient.getVariationName({
+  experimentKey: 'dd_surfacing_full_example_config_experiment_test',
+  groupExperimentName: 'q3_fy22_docs_disco_experiment_group_test',
+  experimentContainer: '#full-config-example',
+}).then((variation) => {
+  if (variation === 'treatment') {
+    $('#full-config-example').css('visibility', 'visible');
+    $('#full-config-example').click(function () {
+      window.AnalyticsClient.trackAction('docs-full-config-example-link', {
+        link: this.id,
```
Comment (rsylvian) [inline src/js/experiments/surfaceFullExampleConfig.js:11]: Same as ma previous comment

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (tshen815) [reply jekyll/_cci2/language-javascript-experimental.md:14]: This guy breaks it down a bit more @ccgeller https://ozzieliu.com/2016/04/26/writing-liquid-template-in-markdown-with-jekyll/

File: src/js/experiments/videoTutorials.js [hunk 19]
```diff
@@ -1,47 +1,43 @@
...
-    if (variation === 'treatment') {
-      $('.main-nav-item[data-section=video-tutorials]').show();
+// https://app.optimizely.com/v2/projects/16812830475/experiments/20790151733/variations
+window.OptimizelyClient.getVariationName({
+  experimentKey: 'dd_add_video_tab_to_docs_test',
+  groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+  experimentContainer: '.main-nav-item[data-section=video-tutorials]',
+}).then((variation) => {
+  if (variation === 'treatment') {
+    $('.main-nav-item[data-section=video-tutorials]').show();
```
Comment (rsylvian) [inline src/js/experiments/videoTutorials.js:8]: same as my previous two comments. This all could be in the same PR btw 😊

File: src/js/site/main.js [hunk 20]
```diff
@@ -6,8 +6,10 @@ hljs.initLineNumbersOnLoad({
   singleLine: true,
 });
 
-const showEvents = ['mouseover', 'hover', 'mouseenter', 'focus'];
-const hideEvents = ['mouseout', 'mouseleave', 'blur'];
+const SHOW_EVENTS = ['mouseover', 'hover', 'mouseenter', 'focus'];
+const HIDE_EVENTS = ['mouseout', 'mouseleave', 'blur'];
+const HEADER_TAGS =
+  'article h1, article h2, article h3, article h4, article h5, article h6';
```
Comment (rsylvian) [inline src/js/site/main.js:12]: 👏

File: src/js/site/sidebar.js [hunk 21]
```diff
@@ -113,13 +111,18 @@ import { isElementInViewport } from '../utils';
  * highlight it.
  *
  * */
-export function highlightTocOnScroll() {
+export function highlightTocOnScroll(headings) {
+  let isExperiment = true;
+  if (!headings) {
+    headings = document.querySelectorAll('h2, h3, h4, h5, h6');
+    isExperiment = false;
+  }
```
Comment (rsylvian) [inline src/js/site/sidebar.js:119]: I am having issues understanding why not having headings means we are in an Experiment?

File: jekyll/index.html (hunk 14 above)
Comment (tshen815) [reply jekyll/index.html:44]: gotcha, most of the hacks Im coming across seems to involve using the `{% raw %}` tag that youre already using

File: src/js/experiments/languageGuides.js (hunk 15 above)
Comment (nramabad) [reply src/js/experiments/languageGuides.js:19]: not intentional. i'll remove it.

File: src/js/site/sidebar.js (hunk 21 above)
Comment (nramabad) [reply src/js/site/sidebar.js:119]: i'm open to other ways of passing this information down but headings is only passed down if we're in treatment. you can see this logic in `experiments/languageGuides.js` and `app.js`

Comment (nramabad) [reply src/js/site/sidebar.js:119]: But yeah otherwise, I can see how these feel rather unrelated/unintuitive. I had that concern while writing it.

**Review by chantelle11 — COMMENTED** [review]
When I navigate between the homepage the the tour I get the green bar flickering
greenbarflicker2

**Review by chantelle11 — COMMENTED** [review]
No tool tip on read time. When I hover over the 50min there is no tooltip about 'Read time'
Screen Shot 2022-01-12 at 8 27 40 PM

**Review by chantelle11 — COMMENTED** [review]
I am very far back from section b but it says I am in section c. This happens when I scroll backwards from c into b
Screen Shot 2022-01-12 at 8 37 15 PM

**Comment by nramabad** [comment]
> No tool tip on read time. When I hover over the 50min there is no tooltip about 'Read time' Screen Shot 2022-01-12 at 8 27 40 PM

I don't think we want a tooltip. @danbeer correct me if I'm wrong but it wasn't in the mocks. there's also questions about the copy so the badge is a WIP.

> I am very far back from section b but it says I am in section c. This happens when I scroll backwards from c into b Screen Shot 2022-01-12 at 8 31 46 PM

I don't think this is unique to my PR?

> When I navigate between the homepage the the tour I get the green bar flickering greenbarflicker2

I'm not sure I'm able to reproduce this.

File: src/js/site/sidebar.js (hunk 21 above)
Comment (rsylvian) [reply src/js/site/sidebar.js:119]: What I don't understand is how a global function that is used on every page is related to your experiment? Couldn't we at least gate based on the URL?

Comment (nramabad) [reply src/js/site/sidebar.js:119]: We can gate based on URL.

The reason why it's dependent is because the treatment and control content have different sections hence different ToC.

Hence this highlight function has to ,at the very least, wait until the experiment JS runs.

**Comment by chantelle11** [comment]
1. (tool tip) Oh sorry I just assumed for it to have a tool tip since that is the new design. My bad!
2. (scrolling) You are right it is not unique to PR, should we make a new ticket for it? I just noticed it while testing PR
3. (green line) It doesn't seem to happen every time but if I keep going back and forth it will show up
greenflicker3

greenflicker4

File: src/js/site/sidebar.js (hunk 21 above)
Comment (nramabad) [reply src/js/site/sidebar.js]: this is done.

File: src/js/experiments/languageGuides.js [hunk 22]
```diff
@@ -0,0 +1,42 @@
+import { reconstructToC, highlightTocOnScrollOnce } from '../site/toc';
+import { displayBlockElement, displayInitialElement } from '../utils';
+
+// https://app.optimizely.com/v2/projects/16812830475/experiments/20872380274/variations
+export default () =>
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_add_video_tab_to_docs_test',
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.treatment',
```
Comment (rsylvian) [inline src/js/experiments/languageGuides.js:9]: This could be potentially dangerous to have a container called just `treatment`. Can we make it more specific to that experiment, please?

File: jekyll/_includes/no-script.html (hunk 13 above)
Comment (rsylvian) [reply jekyll/_includes/no-script.html:2]: Let's add a comment about your great explanation in the file, please 🙏

File: jekyll/_includes/snippets/language-guided-tour-cards.md (hunk 12 above)
Comment (rsylvian) [reply jekyll/_includes/snippets/language-guided-tour-cards.md:10]: nit: Should we rename the directory? We usually use `snippets` for everything that has to do with code snippets and that might be confusing

File: src/js/experiments/languageGuides.js (hunk 7 above)
Comment (rsylvian) [reply src/js/experiments/languageGuides.js:7]: Let's not forget about that please :)

Comment (nramabad) [reply src/js/experiments/languageGuides.js]: haha ofc not

File: jekyll/_includes/no-script.html [hunk 23]
```diff
@@ -0,0 +1,8 @@
+<!--
+    Since by default pages in experiment display none both control and treatment
+    content. It usually depends on JS to display the correct content but in the
+    case of no JavaScript, this noscript tag defaults to showing control content.
+--->
```
Comment (tshen815) [inline jekyll/_includes/no-script.html:5]: 🙏

File: src/js/app.js [hunk 24]
```diff
@@ -22,5 +22,17 @@ $(() => {
...
   services.progressbar.init();
-  import(/* webpackPrefetch: true */ './experiments'); // imports all experiments
+
+  const isGatedPath =
+    ['/docs/', '/docs/2.0/'].includes(window.location.pathname) ||
+    ['-preview/', 'view/2.0/'].includes(window.location.pathname.slice?.(-9)) ||
+    /language-(javascript|python)/gm.test(window.location.pathname);
+  if (!isGatedPath) {
+    site.toc.highlightTocOnScrollOnce();
+  }
```
Comment (tshen815) [inline src/js/app.js:32]: could we add a quick comment here on the purpose of this?

**Review by tshen815 — APPROVED** [review]
Looks great in both preview and locally, approving now but would love some other eyes on it as well as its a big boi PR 🎉

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (devinmarieb) [reply jekyll/_cci2/language-javascript-experimental.md:14]: I'm going to ask about this particular thing again. I was curious so I pulled down the branch and deleted the raw tags and it is fine. What browser/environment is showing it as a code block? I was curious because if this was breaking here it would likely be breaking in other places.

**Review by devinmarieb — APPROVED** [review]
Thanks!

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (nramabad) [reply jekyll/_cci2/language-javascript-experimental.md:14]: @devinmarieb I didn't experience it consistently but only sometimes when it recompiled.

File: src/js/app.js (hunk 24 above)
Comment (nramabad) [reply src/js/app.js]: done.

File: jekyll/_cci2/language-javascript-experimental.md (hunk 1 above)
Comment (devinmarieb) [reply jekyll/_cci2/language-javascript-experimental.md:14]: @nramabad Curious. Might be good to do some kind of audit to see if other pages have this issue because I imagine if it's happening here it's likely happening on other pages as well 🤷‍♀️

File: jekyll/_includes/snippets/language-guided-tour-cards.md [hunk 25]
```diff
@@ -7,7 +7,7 @@
 
 <div id="completion-time" class="card">
 <img alt="Completion Time" src="{{ site.baseurl }}/assets/img/docs/duration.svg">
```
Comment (chantelle11) [inline jekyll/_includes/snippets/language-guided-tour-cards.md:9]: was this icon to be updated as well to not be a clock?

**Review by chantelle11 — APPROVED** [review]
Great job on this! 🥰🥳🌺

**Review by rsylvian — APPROVED** [review]
